### ISSUE TITLE
JS: properties on class and function declarations are not type checked

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#16239]
**TypeScript Version:**  2.3.4

**Code**

```ts
class Foo {}
Foo.bar = 1; // expected error, but got none

const Foo2 = () => {};
Foo2.bar = 1; // error

function FooBar() {}
FooBar.bar = 1; // expected error, but got none

```

This errors is TS files, but not in JS files where `checkJs` is enabled.

**Comment by OliverJAsh** [comment]
Interestingly, the expected/missing errors do appear when the example is wrapped in a block:

``` ts
class Foo {}
Foo.bar = 1; // error

const Foo2 = () => {};
Foo2.bar = 1; // error

function FooBar() {}
FooBar.bar = 1; // error

const FooBar2 = () => {};
FooBar2.bar = 1; // error
```

**Comment by DanielRosenwasser** [comment]
We're intentionally lax about this because there's no such thing as namespace merging (i.e. a function/class/namespace followed by a namespace) in JS. Tacking properties onto values is pretty common for JS users, so I believe we allow it for top-level declarations.

As a matter of fact, this is something we're considering in TS itself. Check out #15868.

**Comment by DanielRosenwasser** [comment]
Actually, the second part about only doing this for top-level declarations is not true - we should not issue an error even in the block scope. You'll see the error go away if you move the statements into a function body. @mhegazy mentioned that that should be a bug.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
